src/H5Tconv_complex.c

```diff
@@ -0,0 +1,2323 @@
+         * complicated. We'll do all the conversion stuff assuming
+         * little endian and then we'll fix the order at the end.
+         */
+        if (H5T_ORDER_BE == src_atomic.order) {
+            uint8_t *cur_part = s;
+            /* Swap real part of complex number element */
+            for (size_t j = 0; j < src_part_size / 2; j++)
+                H5_SWAP_BYTES(cur_part, j, src_part_size - (j + 1));
+            /* Swap imaginary part of complex number element */
+            cur_part += src_part_size;
+            for (size_t j = 0; j < src_part_size / 2; j++)
+                H5_SWAP_BYTES(cur_part, j, src_part_size - (j + 1));
+        }
+        else if (H5T_ORDER_VAX == src_atomic.order)
+            HGOTO_ERROR(H5E_DATATYPE, H5E_UNSUPPORTED, FAIL,
+                        "VAX byte ordering is unsupported for complex number type conversions");
+
+        /* Check for special cases: +0, -0, +Inf, -Inf, NaN */
+        realval_type = H5T__conv_float_find_special(s, &src_atomic, NULL);
+        imagval_type = H5T__conv_float_find_special(s + (src_p->shared->size / 2), &src_atomic, NULL);
+
+        real_zero = (realval_type == H5T_CONV_FLOAT_SPECVAL_POSZERO ||
+                     realval_type == H5T_CONV_FLOAT_SPECVAL_NEGZERO);
+        imag_zero = (imagval_type == H5T_CONV_FLOAT_SPECVAL_POSZERO ||
+                     imagval_type == H5T_CONV_FLOAT_SPECVAL_NEGZERO);
+        real_except =
+            (realval_type == H5T_CONV_FLOAT_SPECVAL_POSINF || realval_type == H5T_CONV_FLOAT_SPECVAL_NEGINF ||
+             realval_type == H5T_CONV_FLOAT_SPECVAL_NAN);
+        imag_except =
+            (imagval_type == H5T_CONV_FLOAT_SPECVAL_POSINF || imagval_type == H5T_CONV_FLOAT_SPECVAL_NEGINF ||
+             imagval_type == H5T_CONV_FLOAT_SPECVAL_NAN);
+
+        /* A complex number is zero if both parts are +/-0 */
+        if (real_zero && imag_zero) {
+            H5T__bit_copy(d, dst_atomic.u.f.sign, s, src_atomic.u.f.sign, (size_t)1);
+            H5T__bit_copy(d + dst_part_size, dst_atomic.u.f.sign, s + src_part_size, src_atomic.u.f.sign,
+                          (size_t)1);
+            H5T__bit_set(d, dst_atomic.u.f.epos, dst_atomic.u.f.esize, false);
+            H5T__bit_set(d + dst_part_size, dst_atomic.u.f.epos, dst_atomic.u.f.esize, false);
+            H5T__bit_set(d, dst_atomic.u.f.mpos, dst_atomic.u.f.msize, false);
+            H5T__bit_set(d + dst_part_size, dst_atomic.u.f.mpos, dst_atomic.u.f.msize, false);
+            goto padding;
+        }
+        else if (real_except || imag_except) {
+            except_ret = H5T_CONV_UNHANDLED;
+
+            /* If user's exception handler is present, use it */
+            if (conv_ctx->u.conv.cb_struct.func) {
+                H5T_conv_except_t except_type; /* type of conversion exception that occurred */
+
+                /* reverse source buffer order first */
+                H5T__reverse_order(src_rev, s, src_p);
```

Comment:
Why do you always reverse the order?

Reply:
This just follows the logic in all the other conversion functions. They all convert the values to little-endian first to make the loops less complicated and then at the point that an exception is thrown the buffer is reversed if necessary before handing it to the exception callback function. If the data is already little-endian this function just copies `s` into `src_rev` (though that's not really necessary and could be optimized).

Reply:
OK, I see that, but why are we always assuming that?   It doesn't necessarily match the datatype IDs.   Seems like an existing bug that would show up with exception handing callbacks on big-endian machines.

Reply:
This would just be putting the value buffer back as it was at the start of the loop before initial byte-swapping, so I'm not sure where the bug would be at.

Reply:
OK, yes, I agree with you.  The asymmetry is annoying: on non-LE types, it byte swaps, but is memcpy for LE types.  Very weird.